### ISSUE TITLE
MLContextOptions.deviceType seems unnecessary outside of conformance testing

### ISSUE THREAD

**Issue by mwyrzykowski** [webmachinelearning/webnn#749]
WebKit has a few objections to MLContextOptions.deviceType:

### MLContextOptions.deviceType is currently unimplementable currently via CoreML
CoreML is the only public framework on macOS for running computations on the NPU / ANE. The reason as pointed out https://github.com/webmachinelearning/webnn/issues/623#issuecomment-2025767956 and possibly elsewhere, CoreML allows you to specify MLComputeUnitsCPUOnly, MLComputeUnitsCPUAndGPU, MLComputeUnitsAll, and MLComputeUnitsCPUAndNeuralEngine, but not MLComputeUnitsNeuralEngine only for instance. Even assuming we could request CoreML to add support for only the ANE, that support would be limited to a future version of macOS and therefore could only be supported on currently unreleased versions of macOS.

Also note, MLComputeUnitsCPUAndNeuralEngine, is only available on macOS 13.0 and later and I'm not sure if some browser vendors want to support earlier versions of macOS.

### MLContextOptions.deviceType, as currently specified, would lead to additional fragmentation, works on some devices, fails on others
Under [Section 4](https://www.w3.org/TR/webnn/#privacy) we see the wording:
> if this type cannot be satisfied, an "[OperationError](https://webidl.spec.whatwg.org/#operationerror)" [DOMException](https://webidl.spec.whatwg.org/#idl-DOMException) is thrown,

It would be preferable to fallback instead of failing, otherwise a website which specifies NPU and runs on a variety of NPUs, but maybe not the NPU of a certain device, the website would not run if the author does not handle fallback. This seems unnecessary, when the browser could fallback to the GPU or CPU themselves.

### The browser has better insight into workloads than the website author
It was pointed out in https://github.com/webmachinelearning/webnn/pull/322 that a website author may want their models to not run on the GPU so they don't impact WebGPU rendering performance. While a good idea in theory, the browser implementation and native frameworks have much greater insight into the graphics chip utilization and the most appropriate resource to run a model on without impacting target frameworks.

The answer to this question, where to run the model, can not be easily decided by the website as the website does not know the underlying hardware it is running on due to privacy reasons and possibly other reasons. Additionally, while browser vendor can trivially compute the amount of time a GPUCommandBuffer takes to submit from WebGPU, such timing is not available to the website. For example, if all the GPUCommandBuffers in the WebGPU workload complete in <3ms on a Mac Studio, and the website calls requestAnimationFrame every ~16ms, then restricting the WebNN model from running on the GPU seems unnecessary. However on a less powerful device, like an iPhone 12 mini, if the WebGPU workload is falling behind the requested update rate, the browser implementation may move the model processing to the device least in use. To the website, a browser implementation may present both devices as having the same capabilities for privacy reasons, so there is no way for the website to make the correct decision which device to run on.

Often with graphics intensive WebGPU applications, we find the GPU is still the least utilized processor as the CPU overhead can be high. Long running shaders or complex compute kernels are two exceptions. In any case this is device dependent and not something a website author can easily predict.

To summarize it seems MLDeviceType is useful for testing in browser implementations, but seems counter-productive for framework authors and general websites / wasm applications.

**Comment by zolkis** [comment]
To summarize the points above:
- `deviceType` option is hard to standardize because of the heterogeneity of the compute units across various platforms, and even across their versions.
- fallback is preferable instead of failing, and implementations/the underlying platforms should determine the fallback type based on runtime information.
- implementations, browser, OS have better grasp of the system/compute/runtime/apps state then websites, therefore control should be relished to them.

Some of these points were mentioned in #696 (adding "npu" to [`deviceType`](https://webmachinelearning.github.io/webnn/#dom-mlcontextoptions-devicetype)) and discussed in #623 (support for NPU, QDQ, minimum op set).

We had a short discussion on this during the last [call](https://www.w3.org/2024/10/17-webmachinelearning-minutes.html) (section 4). Some questions for further consideration:
- are there use cases when web applications are part of larger solutions and would like to affect the type of AI accelerators used with the model selected (by them or by their user). For reasons above, and also because of [security issues](https://webmachinelearning.github.io/webnn/#issue-e9d96711), this could possibly be only a _hint_.
- If there are such use cases, can we defer them to non-breaking future changes in the API or its behavior?
- How do you propose to change the API: would we keep/change the `powerPreference` [ContextOptions](https://webmachinelearning.github.io/webnn/#api-mlcontextoptions) and remove `deviceType` like in #322?
- Or, is it possible we keep / standardize better the `deviceType` (name can also change) to better reflect the use case/developer intentions, without getting into possible fragmentation issues? What about calling it `devicePreference` to emphasize this is a preference, not a required option? Handle it as a _hint_, and map it to the real underlying compute structure the best way the implementation/underlying platform decides to do? For instance, `"npu"` preference might mean NPU on one platform, NPU+CPU on another, or the nearest fallback (which can be GPU). We might add an informative mapping array to the spec for known frameworks.
- Or, remove both and rethink if we could possibly pass any other possible extra (standardized) information about the model/graph that might help the underlying impl/platform figure out the best how to run the graph, and/or express developer intentions for compute preferences?

**Comment by mwyrzykowski** [comment]
I wonder if we could achieve some of the use cases by generalizing the concept and abstracting it a little?

E.g. (not a formal proposal),

```
interface ML {
  Promise<MLContext> createContext(optional MLContextOptions options = {});
  Promise<MLContext> createContext(GPUDevice gpuDevice);
};
```
becomes:
```
interface ML {
  Promise<MLContext> createContext(optional MLContextOptions options = {});
  Promise<MLContext> createContext(GPUDevice gpuDevice);
  
  record<USVString, MLOpSupportLimits> opSupportLimitsPerAdapter();
};
```
where an implementation may choose to implement `opSupportLimitsPerAdapter` in a way such that it returns a dictionary of `MLOpSupportLimits` for each compute device on a system (CPU, GPU, NPU).

Alternatively an implementation may choose to return a single value in the dictionary. Furthermore, an implementation may choose to return values which preserve battery life, offer higher performance at the expense of battery life, avoid impacting other NPU or GPU operations, etc.

Then, `MLContextOptions` can become:
```
dictionary MLContextOptions {
  MLOpSupportLimits requiredLimits = {};
};
```

where `requiredLimits` corresponds to a value returned by `opSupportLimitsPerAdapter` and indicates to the implementation which limits will be required during the lifetime of the MLContext.

Then the website/app author / framework developer chooses the set of `MLOpSupportLimits` that maps most directly to their needs and passes it to the `ML.createContext` call.

I think that would support the DirectML use case where an implementation needs to know which device the tensors will run on while supporting CoreML which doesn't allow requiring the NPU. Additionally, it can preserve privacy as an implementation can decide the granularity of the information returned from `opSupportLimitsPerAdapter`

**Comment by anssiko** [comment]
[Discussed on 2024-10-31](https://www.w3.org/2024/10/31-webmachinelearning-minutes.html#f739): proposed next step to write an explainer.

@zolkis do you want to take the first stab at the explainer? Synthesize design proposals and considered alternatives from this issue and other issues, from our telcons, use case discussions. Basically document the paths we've explored as a group into a document. We'd then evolve the doc toward a proposed solution as a group and could offer it to e.g. privacy people for their feedback.

Here are some pointers (not a complete list!):

- Issues labeled with https://github.com/webmachinelearning/webnn/labels/device%20selection
- https://www.w3.org/2024/10/31-webmachinelearning-minutes.html#f739 (MLOpSupportLimits proposal)
- https://www.w3.org/2024/10/17-webmachinelearning-minutes.html#b982 (use cases)
- https://lists.w3.org/Archives/Public/www-archive/2024Sep/att-0006/MLDeviceType.pdf (TPAC presentation)
- https://www.w3.org/2024/09/23-webmachinelearning-minutes.html#b5c1 (TPAC discussion)
- https://www.w3.org/2024/09/05-webmachinelearning-minutes.html#t11 (frameworks)
- https://www.w3.org/2024/08/22-webmachinelearning-minutes.html#t02 (WebGPU, CoreML, use cases)
- https://github.com/webmachinelearning/webnn/issues/623 (options for extending DeviceType, use cases: compute offloading, op fallback)
- https://github.com/webmachinelearning/webnn/pull/696 (MLDeviceType npu PR, see comments)
- https://gpuweb.github.io/gpuweb/#gpusupportedlimits (GPUSupportedLimits precedent)

**Comment by mwyrzykowski** [comment]
Resolved by #809 as discussed in today's call
